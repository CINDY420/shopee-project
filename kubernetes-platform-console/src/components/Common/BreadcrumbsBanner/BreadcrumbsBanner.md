
```jsx
import * as React from 'react'
import { Router } from 'react-router'

import history from 'helpers/history'

const CrumbType  = {
  REQUESTS: 'requests',
  REQUEST: 'request'
}

const routeCrumbsTree = [
  {
    route: '/',
    extendedCrumbs: CrumbType.REQUESTS,
    children: [
      {
        route: '/request',
        extendedCrumbs: CrumbType.REQUEST
      }
    ]
  }
]

const crumbLookup = {
  [CrumbType.REQUESTS]: {
    onClick: data => {
      history.push('/')
    }
  }
}

const crumbResourceLookup = {
  [CrumbType.REQUESTS]: { name: 'All requests' },
  [CrumbType.REQUEST]: { name: 'Request detail' }
}

const uselessFunction = () => {}

<Router history={history}>
  <BreadcrumbsBanner
    crumbLookup={crumbLookup}
    crumbResourceLookup={crumbResourceLookup}
    routeCrumbsTree={routeCrumbsTree}
  />
</Router>
```


