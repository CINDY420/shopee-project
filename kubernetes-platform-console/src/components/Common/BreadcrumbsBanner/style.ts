import styled from 'styled-components'

interface IName {
  isActive?: boolean
}

export const Root = styled.div`
  width: 100%;
  padding: 1em 0;
  font-size: 12px;
  box-sizing: border-box;
  height: 45px;
`

export const Crumb = styled.div`
  display: flex;
  align-items: center;
`

export const Name = styled.div`
  font-weight: normal;
  cursor: ${(props: IName) => (props.isActive ? 'pointer' : 'default')};
  color: ${(props: IName) => (props.isActive ? 'rgba(0, 0, 0, 0.45)' : '#4a4a4a')};

  &:hover {
    color: ${(props: IName) => (props.isActive ? '#4a90e2' : '#4a4a4a')};
  }
`
