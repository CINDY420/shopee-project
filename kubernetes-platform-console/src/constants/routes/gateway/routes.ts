import { GATEWAY_ID, TENANT_ID, PROJECT_ID, DOMAIN_GROUP_ID, DOMAIN_ID, CLUSTER_ID } from '../identifier'
import { TENANT, PROJECT, RESOURCE, CLUSTER } from '../name'
import { buildRoute } from '../routes'

export const GATEWAYS = '/gateways'

export const GATEWAY_SERVICES = `${GATEWAYS}/Services`
export const GATEWAY_LOAD_BALANCE = `${GATEWAYS}/Load Balance`
export const GATEWAY_INGRESS = `${GATEWAYS}/Ingress`

export const GATEWAY = `${GATEWAYS}/:${GATEWAY_ID}`
export const GATEWAY_SERVICES_GROUP_DETAIL = `${GATEWAY_SERVICES}/${TENANT}`
export const GATEWAY_SERVICES_PROJECT_DETAIL = `${GATEWAY_SERVICES}/${PROJECT}`
export const GATEWAY_INGRESS_LIST = `${GATEWAY_INGRESS}/${CLUSTER}`
export const GATEWAY_DOMAIN_GROUP = `${GATEWAYS}/:${GATEWAY_ID}/domainGroups/:${DOMAIN_GROUP_ID}`
export const GATEWAY_DOMAIN = `${GATEWAY_DOMAIN_GROUP}/clusters/:${CLUSTER_ID}/domains/:${DOMAIN_ID}`

export const GATEWAY_LOAD_BALANCE_RESOURCE = `${GATEWAY_LOAD_BALANCE}/${RESOURCE}`
export const GATEWAY_LOAD_BALANCE_GIT_TEMPLATE_RENDER = `${GATEWAY_LOAD_BALANCE}/resources/Git Template Render`

export const LOAD_BALANCE_DOMAIN_GROUP = `${GATEWAYS}/Load Balance/domainGroups`
export const GATEWAY_LOAD_BALANCE_DOMAIN_GROUP = `${LOAD_BALANCE_DOMAIN_GROUP}/:${DOMAIN_GROUP_ID}`
export const GATEWAY_LOAD_BALANCE_DOMAIN = `${GATEWAY_LOAD_BALANCE_DOMAIN_GROUP}/Clusters/:${CLUSTER_ID}/Domains/:${DOMAIN_ID}`

export const buildGatewayGroupRoute = ({ gatewayType, tenantId }) =>
  buildRoute(GATEWAY_SERVICES_GROUP_DETAIL, {
    [GATEWAY_ID]: gatewayType,
    [TENANT_ID]: tenantId
  })

export const buildGatewayProjectRoute = ({ gatewayType, tenantId, projectName }) =>
  buildRoute(GATEWAY_SERVICES_PROJECT_DETAIL, {
    [GATEWAY_ID]: gatewayType,
    [TENANT_ID]: tenantId,
    [PROJECT_ID]: projectName
  })

export const buildLoadBalanceDomainGroupRoute = ({ domainGroupId }) =>
  buildRoute(GATEWAY_LOAD_BALANCE_DOMAIN_GROUP, {
    [DOMAIN_GROUP_ID]: domainGroupId
  })

export const buildGatewayLoadBalanceDomainRoute = ({ domainGroupName, domainName, clusterName }) =>
  buildRoute(GATEWAY_LOAD_BALANCE_DOMAIN, {
    [GATEWAY_ID]: 'Load Balance',
    [DOMAIN_GROUP_ID]: domainGroupName,
    [DOMAIN_ID]: domainName,
    [CLUSTER_ID]: clusterName
  })
