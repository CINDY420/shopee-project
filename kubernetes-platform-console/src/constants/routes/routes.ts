import {
  TENANT,
  PROJECT,
  APPLICATION,
  POD,
  INGRESS,
  SERVICE,
  DEPLOYMENT,
  PIPELINE_RUNNING,
  PIPELINE_HISTORIES,
  EDITPIPELINE
} from './name'
import {
  TENANT_ID,
  PROJECT_ID,
  APPLICATION_ID,
  DEPLOYMENT_ID,
  CLUSTER_ID,
  POD_ID,
  NODE_ID,
  PIPELINE_NAME,
  PIPELINE_RUNNING_ID,
  PROFILE_ID
} from './identifier'

export const ROOT = '/'
export const LOGIN = '/login'

export const PLATFORM = '/platform'
export const CLUSTERS = PLATFORM + '/clusters'
export const CLUSTER_DETAIL = CLUSTERS + '/:clusterId'
export const NODE_DETAIL: string = CLUSTER_DETAIL + '/nodes/:nodeId'
export const PLATFORMS = PLATFORM + '/platforms'
export const TENANTS = PLATFORM + '/tenants'
export const RESOURCE = PLATFORM + '/resource'

export const APPLICATIONS = '/applications'
export const INGRESSES = '/ingresses'
export const PIPELINES = '/pipelines'
export const PIPELINES_LIST = PIPELINES + '/tenants'
export const PIPELINE_OVERVIEW = `${PIPELINES}/${TENANT}`
export const CREATE_PIPELINE = PIPELINE_OVERVIEW + '/create'
export const EDIT_PIPELINE = `${PIPELINES}/${EDITPIPELINE}`
export const PIPELINE_DETAIL = `${PIPELINES}/${PIPELINE_HISTORIES}`
export const PIPELINE_RUN = `${PIPELINES}/${PIPELINE_RUNNING}`
export const RELEASE_FREEZES = `${PLATFORM}/releaseFreezes`

export const DEFAULT_ROUTE: string = APPLICATIONS

export const TENANT_DETAIL = `${APPLICATIONS}/${TENANT}`
export const PROJECT_DETAIL = `${APPLICATIONS}/${PROJECT}`
export const PROJECT_CREATE = `${APPLICATIONS}/${TENANT}/create`
export const PROJECT_EDIT = `${APPLICATIONS}/${PROJECT}/edit`
export const APPLICATION_DETAIL = `${APPLICATIONS}/${APPLICATION}`
export const INGRESS_DETAIL = `${APPLICATIONS}/${INGRESS}`
export const SERVICE_DETAIL = `${APPLICATIONS}/${SERVICE}`
export const POD_DETAIL = `${APPLICATIONS}/${POD}`
export const POD_FLAME_GRAPH = `${POD_DETAIL}/profiles/:${PROFILE_ID}/flameGraph`

export const DEPLOYMENT_DETAIL = `${APPLICATIONS}/${DEPLOYMENT}`
export const APPLICATIONS_NOT_FOUND: string = APPLICATIONS + '/404'

export const REDIRECT_QUERY_KEY = 'redirect'

export const APPROVALS = '/approvals'
export const APPROVAL_ID = 'approvalId'
export const APPROVAL_DETAIL = `${APPROVALS}/:${APPROVAL_ID}`

export const REQUESTS = '/requests'
export const REQUESTS_ID = 'requestId'
export const REQUESTS_DETAIL = `${REQUESTS}/:${REQUESTS_ID}`

export const OPERATION_LOGS = '/operationLogs'

export const GATEWAYS = '/gateways'

export const ACCESS_REQUEST = '/accessRequest'

export const RBAC_ACTION_KEY = 'RBACActionType'

export const TICKET_CENTER_KEY = 'ticketsCenter'
export const TICKET_CENTER = `/${TICKET_CENTER_KEY}`

export const PENDING_MY_ACTION = 'pendingMyAction'
export const MY_REQUESTS = 'myRequests'
export const PENDING_MY_ACTION_LIST = `${TICKET_CENTER}/${PENDING_MY_ACTION}`
export const MY_REQUESTS_LIST = `${TICKET_CENTER}/${MY_REQUESTS}`

export const TICKET_LIST_TYPE = 'ticketListType'
export const TICKET_LIST_KEY = `${TICKET_CENTER_KEY}/:${TICKET_LIST_TYPE}`
export const TICKET_LIST = `${TICKET_CENTER}/:${TICKET_LIST_TYPE}`

export const TICKET_ID = 'ticketId'
export const TICKET_DETAIL_KEY = 'ticketDetail'
export const TICKET_DETAIL = `${TICKET_LIST}/${TICKET_DETAIL_KEY}/:${TICKET_ID}`

export const buildRoute = (pathTemplate: string, configs: { [id: string]: string }): string => {
  let result = pathTemplate
  Object.entries(configs).forEach(([id, str]) => {
    result = result.replace(`:${id}`, str)
  })

  return result
}

export const buildGroupDetailRoute = ({ tenantId }) =>
  buildRoute(TENANT_DETAIL, {
    [TENANT_ID]: tenantId
  })

export const buildProjectDetailRoute = ({ tenantId, projectName }) =>
  buildRoute(PROJECT_DETAIL, {
    [TENANT_ID]: tenantId,
    [PROJECT_ID]: projectName
  })

export const buildApplicationDetailRoute = ({ tenantId, projectName, applicationName }) =>
  buildRoute(APPLICATION_DETAIL, {
    [TENANT_ID]: tenantId,
    [PROJECT_ID]: projectName,
    [APPLICATION_ID]: applicationName
  })

export const buildDeployDetailRoute = ({ tenantId, projectName, applicationName, deployName, clusterId }) =>
  buildRoute(DEPLOYMENT_DETAIL, {
    [TENANT_ID]: tenantId,
    [PROJECT_ID]: projectName,
    [APPLICATION_ID]: applicationName,
    [DEPLOYMENT_ID]: deployName,
    [CLUSTER_ID]: clusterId
  })

export const buildPodDetailRoute = ({ tenantId, projectName, applicationName, deployName, clusterName, podName }) =>
  buildRoute(POD_DETAIL, {
    [TENANT_ID]: tenantId,
    [PROJECT_ID]: projectName,
    [APPLICATION_ID]: applicationName,
    [DEPLOYMENT_ID]: deployName,
    [CLUSTER_ID]: clusterName,
    [POD_ID]: podName
  })

export const buildNodeDetailRoute = ({ clusterName, nodeName }) =>
  buildRoute(NODE_DETAIL, {
    [CLUSTER_ID]: clusterName,
    [NODE_ID]: nodeName
  })

export const buildPipelineRoute = ({ tenantId }) =>
  buildRoute(PIPELINES, {
    [TENANT_ID]: tenantId
  })

export const buildPipelineDetailRoute = ({ tenantId, pipelineName }) =>
  buildRoute(PIPELINE_DETAIL, {
    [TENANT_ID]: tenantId,
    [PIPELINE_NAME]: pipelineName
  })

export const buildPipelineRunRoute = ({ tenantId, pipelineName, runId }) =>
  buildRoute(PIPELINE_RUN, {
    [TENANT_ID]: tenantId,
    [PIPELINE_NAME]: pipelineName,
    [PIPELINE_RUNNING_ID]: runId
  })

export const buildFlameGraphRoute = ({
  tenantId,
  projectName,
  applicationName,
  deployName,
  clusterId,
  podName,
  profileId
}) =>
  buildRoute(POD_FLAME_GRAPH, {
    [TENANT_ID]: tenantId,
    [PROJECT_ID]: projectName,
    [APPLICATION_ID]: applicationName,
    [DEPLOYMENT_ID]: deployName,
    [CLUSTER_ID]: clusterId,
    [POD_ID]: podName,
    [PROFILE_ID]: profileId
  })
