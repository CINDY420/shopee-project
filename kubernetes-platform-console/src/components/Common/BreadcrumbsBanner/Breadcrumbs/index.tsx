import * as React from 'react'
import { Breadcrumb } from 'infrad'
import warning from 'warning'

import Collapsed from './Collapsed'
import { StyledBreadcrumb } from './style'

interface IComponentProps {
  className?: string
  children: React.ReactNode
}

interface IBreadcrumbsProps {
  className?: string
  children: React.ReactNode
  component?: string | React.ReactType<IComponentProps>
  itemsAfterCollapse?: number
  itemsBeforeCollapse?: number
  maxItems?: number
  separator?: string | React.ReactNode
}

const Breadcrumbs: React.FC<IBreadcrumbsProps> = props => {
  const {
    maxItems = 8,
    itemsAfterCollapse = 2,
    itemsBeforeCollapse = 2,
    component: Component = 'nav',
    children,
    className,
    separator
  } = props

  const renderItems = (allItems: any[]) => {
    if (itemsAfterCollapse + itemsBeforeCollapse >= maxItems) {
      warning(
        false,
        [
          'Common Component: you have provided an invalid combination of properties to the Breadcrumbs.',
          `itemsAfterCollapse={${itemsAfterCollapse}} + itemsBeforeCollapse={${itemsBeforeCollapse}} >= maxItems={${maxItems}}`
        ].join('\n')
      )

      return allItems
    }

    if (allItems.length > maxItems) {
      const end = allItems.length - itemsAfterCollapse
      return (
        <>
          {allItems.slice(0, itemsBeforeCollapse).map((item, index) => (
            <Breadcrumb.Item key={index}>{item}</Breadcrumb.Item>
          ))}
          <Collapsed>{allItems.slice(itemsBeforeCollapse, end)}</Collapsed>
          {allItems.slice(end).map((item, index) => (
            <Breadcrumb.Item key={index + end}>{item}</Breadcrumb.Item>
          ))}
        </>
      )
    }

    return allItems.map((item, index) => <Breadcrumb.Item key={index}>{item}</Breadcrumb.Item>)
  }

  const allItems = React.Children.toArray(children)

  return (
    <Component className={className}>
      <StyledBreadcrumb separator={separator}>{renderItems(allItems)}</StyledBreadcrumb>
    </Component>
  )
}

export default Breadcrumbs
