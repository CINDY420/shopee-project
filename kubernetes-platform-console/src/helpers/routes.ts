import { Location } from 'history'
import { generatePath } from 'react-router-dom'
import { matchPath } from 'react-router'
import { MonitoringPlatformUrl, LOG_PLATFORM_ENV_URL_MAP } from 'constants/routes/external'

import history from 'helpers/history'
import { stringify } from 'query-string'
import { parseClusterIdWithFte } from 'helpers/cluster'

export const getMatchResourceHandler = (handler: (matches: any, path: string, location: Location) => void) => ({
  handler
})

const LIVE_ENVS = ['live', 'liveish']

interface INameObject {
  name: string
}

interface IGenerateLogPlatformLink {
  clusterId?: string
  env?: string
  cid?: string
  projectName: string
  appName: string
  nodeIP?: string
  podIP?: string
}
interface IGenerateTracingLink {
  tracingUrl: string
  traceId: string
}
interface IGenerateMonitoringPlatformLink {
  tenantName: string
  env: string
  cid: string
  projectName: string
  appName: string
  monitoringClusterName: string
  clusterName: string
}
/**
 * Get pod container route (for the sake of backend hardcode shit)
 */
export const getPodContainerRoute = (
  tenantId: number,
  project: INameObject,
  application: INameObject,
  pod: INameObject,
  containerName: string
) =>
  `tenants/${tenantId}/projects/${project.name}/apps/${application.name}/pods/${pod.name}/containers/${containerName}`

interface IOptions {
  replace?: boolean
  state?: any
}

/**
 * change route
 * @param pattern string  A pattern provided as a path attribute to the Route component.
 * @param params object  Object with corresponding params for the pattern to use. If provided params and path donâ€™t match, an error will be thrown.
 * @param options object  Object containing replace and state.
 */
export const changeRouteWithParams = (
  pattern: string,
  params: { [paramName: string]: string | number | boolean | undefined } = {},
  options: IOptions = {}
) => {
  const { replace = false, state } = options
  const targetPath = generatePath(pattern, params)
  if (replace) {
    history.replace(targetPath, state)
  } else {
    history.push(targetPath, state)
  }
}

/**
 * destruct object from source with route
 * @param source source
 * @param path path
 * @example source: a/1/b/2 , path: a/:aName/b/:bName => {a:1, b:2}
 */
export const destructFromRoute = (source: string, path: string) => {
  const match = matchPath(source, {
    path,
    exact: true,
    strict: false
  })
  const { params }: any = match
  return params
}

/**
 * check if route match path
 * @param route route to match
 * @param path path
 * @returns [match, isExact]
 */
export const checkIfRouteMatchPath = (route: string, path: string): [boolean, boolean] => {
  const match = matchPath(path, { path: route })
  if (match == null) {
    return [false, false]
  }
  return [true, match.isExact]
}

export const generateLogPlatformLink = ({
  clusterId,
  env: passedEnv,
  cid: passedCid,
  projectName,
  appName,
  nodeIP
}: IGenerateLogPlatformLink): string => {
  let env = passedEnv
  let cid = passedCid
  if (clusterId) {
    const { env: parsedEnv, cid: parsedCid } = parseClusterIdWithFte(clusterId)
    env = parsedEnv
    cid = parsedCid
  }

  const lowerCaseEnv = env.toLowerCase()
  const lowerCaseCid = cid.toLowerCase()

  const applicationItems = appName.split('-')
  const logModule = applicationItems.length > 1 ? applicationItems.slice(1).join('-') : undefined

  const logQs = stringify({
    payload: JSON.stringify({
      project: projectName,
      module: logModule,
      env: lowerCaseEnv,
      cid: lowerCaseCid,
      platform: 'PLATFORM_K8S_SZ',
      keyword: nodeIP && `@ip:"${nodeIP}"`
    })
  })
  let logPlatformUrl = LOG_PLATFORM_ENV_URL_MAP[__SERVER_ENV__]
  if (__SERVER_ENV__ === 'live') {
    if (LIVE_ENVS.indexOf(lowerCaseEnv) === -1) {
      logPlatformUrl = LOG_PLATFORM_ENV_URL_MAP.nonlive
    }
  }
  const logPlatformLink = `${logPlatformUrl}?${logQs}`
  return logPlatformLink
}

export const generateTracingPlatformLink = ({ tracingUrl, traceId }: IGenerateTracingLink): string => {
  const tracingPlatformLink = `${tracingUrl}/${traceId}`
  return tracingPlatformLink
}

export const addProtocol = (url: string) => {
  const hasProtocol = /(http|https):\/\/([\w.]+\/?)\S*/.test(url)
  const protocol = location.protocol
  return hasProtocol ? url : `${protocol}//${url}`
}

export const generateMonitoringPlatformLink = ({
  tenantName,
  env,
  cid,
  projectName,
  appName,
  monitoringClusterName,
  clusterName
}: IGenerateMonitoringPlatformLink): string => {
  const lowerCaseEnv = env.toLowerCase()
  const lowerCaseCid = cid.toLowerCase()

  const tenantItems = tenantName.split(' ')
  const parsedClusterName = monitoringClusterName || clusterName.replace(/^kube/, 'k8s')
  const parsedGroup = tenantItems.length ? tenantItems.join('-') : undefined
  const monitorQs = stringify({
    'var-cluster': parsedClusterName,
    'var-group': parsedGroup,
    'var-project': projectName,
    'var-env': lowerCaseEnv,
    'var-module': appName,
    'var-cid': lowerCaseCid,
    orgId: '1' // Specify monitor organization
  })
  const monitoringHref = `${MonitoringPlatformUrl}?${monitorQs}`
  return monitoringHref
}
