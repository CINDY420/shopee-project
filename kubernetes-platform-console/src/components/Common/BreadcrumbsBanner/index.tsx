import * as React from 'react'
import Breadcrumbs from 'components/Common/BreadcrumbsBanner/Breadcrumbs'
import useLocationChange from 'hooks/routes/useLocationChange'
import { checkIfRouteMatchPath } from 'helpers/routes'

import { Root, Crumb, Name } from './style'

interface ICrumb {
  key?: string
  desc?: string
  onClick?: (param: any) => void
  overlay?: (props: any) => React.ReactElement
  isHideResourceName?: boolean
  nameKey?: string
}

interface IRouteCrumbsNode {
  route: string
  children?: Array<IRouteCrumbsNode>
  extendedCrumbs: string | Array<string>
}

type IRouteCrumbsTree = Array<IRouteCrumbsNode>

/**
 * This component is used for creating bread crumbs in K8S system
 */
export interface IBreadcrumbsBannerProps {
  /** Route jump path when user click the crumb */
  crumbLookup: {
    [crumbKey: string]: ICrumb
  }
  /** Current selected crumb names */
  crumbResourceLookup: {
    [crumbKey: string]: {
      [key: string]: any
      name?: string
    }
  }
  /** Route crumbs tree definition */
  routeCrumbsTree: IRouteCrumbsTree
}

type IActiveCrumb = [string, ICrumb]
type IActiveCrumbs = Array<IActiveCrumb>

const findTheExactMatchRouteCrumbsNodePath = (
  pathname,
  routeCrumbsArray: Array<IRouteCrumbsNode>
): Array<IRouteCrumbsNode> => {
  let currentMatchRouteCrumbsNodePath = []

  for (const routeCrumbsNode of routeCrumbsArray) {
    const [match, isExact] = checkIfRouteMatchPath(routeCrumbsNode.route, pathname)

    if (!match) {
      continue
    }

    if (isExact) {
      currentMatchRouteCrumbsNodePath = [routeCrumbsNode]
      break
    }

    if (routeCrumbsNode.children && routeCrumbsNode.children.length) {
      const childPath = findTheExactMatchRouteCrumbsNodePath(pathname, routeCrumbsNode.children)
      if (childPath.length) {
        currentMatchRouteCrumbsNodePath = [routeCrumbsNode, ...childPath]
      } else {
        break
      }
    }
  }

  return currentMatchRouteCrumbsNodePath
}

const BreadcrumbsBanner: React.FC<IBreadcrumbsBannerProps> = ({
  crumbLookup,
  crumbResourceLookup,
  routeCrumbsTree
}) => {
  const [activeCrumbs, setActiveCrumbs] = React.useState<IActiveCrumbs>([])

  const handleLocationChange = React.useCallback(
    location => {
      // TODO, should handle hash change here?
      const { pathname } = location
      const exactMatchRouteCrumbsNodePath = findTheExactMatchRouteCrumbsNodePath(pathname, routeCrumbsTree)
      if (!exactMatchRouteCrumbsNodePath.length) {
        console.error('Oops! can not find match route crumbsNode for ', pathname)
        return
      }

      const activeCrumbKeys = exactMatchRouteCrumbsNodePath.reduce((crumbKeys, node) => {
        return crumbKeys.concat(node.extendedCrumbs)
      }, [])
      const activeCrumbs = activeCrumbKeys.map<IActiveCrumb>(crumbKey => {
        const crumb = crumbLookup[crumbKey]
        if (crumb) {
          return [crumbKey, crumb]
        } else {
          console.error('missing crumb for key: ', crumbKey)
          return [crumbKey, {}]
        }
      })

      setActiveCrumbs(activeCrumbs)
    },
    [routeCrumbsTree, crumbLookup]
  )

  useLocationChange(handleLocationChange)

  return (
    <Root>
      <Breadcrumbs>
        {activeCrumbs.map(
          ([crumbType, { key, desc, onClick, overlay, isHideResourceName, nameKey = 'name' }], index) => {
            const resource = crumbResourceLookup[crumbType]
            const displayName = resource && resource[nameKey]
            const clickable = index !== activeCrumbs.length - 1 && !!onClick

            return displayName || desc ? (
              <Crumb key={key || index}>
                <Name isActive={clickable} onClick={() => clickable && onClick && onClick(resource)}>
                  {desc || null}
                  {isHideResourceName ? '' : displayName}
                </Name>
                {overlay && overlay(resource)}
              </Crumb>
            ) : null
          }
        )}
      </Breadcrumbs>
    </Root>
  )
}

export default BreadcrumbsBanner
