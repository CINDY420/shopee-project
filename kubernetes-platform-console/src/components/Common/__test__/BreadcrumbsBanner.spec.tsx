import * as React from 'react'
import * as handleLocationChange from 'hooks/routes/useLocationChange'
import { render, act, screen, fireEvent } from '@testing-library/react'
import BreadcrumbsBanner, { IBreadcrumbsBannerProps } from '../BreadcrumbsBanner'

describe('BreadcrumbsBanner render UI result', () => {
  it('should display crumb text according to crumbResourceLookup configuration', () => {
    const crumbsLookup: IBreadcrumbsBannerProps['crumbLookup'] = {
      a: {}
    }
    const crumbResourceLookup: IBreadcrumbsBannerProps['crumbResourceLookup'] = {
      a: { name: 'a' }
    }
    const crumbs: IBreadcrumbsBannerProps['routeCrumbsTree'] = [
      {
        route: '/a',
        extendedCrumbs: 'a'
      }
    ]

    const handleLocationChangeMock = jest.spyOn(handleLocationChange, 'default')
    let innerHandleLocationChange = null

    handleLocationChangeMock.mockImplementation(_handleLocationChange => {
      innerHandleLocationChange = _handleLocationChange
    })

    const { container } = render(
      <BreadcrumbsBanner
        crumbLookup={crumbsLookup}
        crumbResourceLookup={crumbResourceLookup}
        routeCrumbsTree={crumbs}
      />
    )
    act(() => {
      innerHandleLocationChange({ pathname: '/a' })
    })

    expect(container).toMatchSnapshot()
  })

  it('should display crumb text with description as prefix according to crumbLookup configuration', () => {
    const crumbsLookup: IBreadcrumbsBannerProps['crumbLookup'] = {
      a: { desc: 'A:' }
    }
    const crumbResourceLookup: IBreadcrumbsBannerProps['crumbResourceLookup'] = {
      a: { name: 'a' }
    }
    const crumbs: IBreadcrumbsBannerProps['routeCrumbsTree'] = [
      {
        route: '/a',
        extendedCrumbs: 'a'
      }
    ]

    const handleLocationChangeMock = jest.spyOn(handleLocationChange, 'default')
    let innerHandleLocationChange = null

    handleLocationChangeMock.mockImplementation(_handleLocationChange => {
      innerHandleLocationChange = _handleLocationChange
    })

    const { container } = render(
      <BreadcrumbsBanner
        crumbLookup={crumbsLookup}
        crumbResourceLookup={crumbResourceLookup}
        routeCrumbsTree={crumbs}
      />
    )
    act(() => {
      innerHandleLocationChange({ pathname: '/a' })
    })

    expect(container).toMatchSnapshot()
  })

  it('should render children crumb according to routeCrumbsTree configuration', () => {
    const crumbsLookup: IBreadcrumbsBannerProps['crumbLookup'] = {
      a: {},
      b: {},
      c: {}
    }
    const crumbResourceLookup: IBreadcrumbsBannerProps['crumbResourceLookup'] = {
      a: { name: 'a' },
      b: { name: 'b' },
      c: { name: 'c' }
    }
    const crumbs: IBreadcrumbsBannerProps['routeCrumbsTree'] = [
      {
        route: '/a',
        extendedCrumbs: 'a',
        children: [
          {
            route: '/a/b',
            extendedCrumbs: 'b',
            children: [
              {
                route: '/a/b/c',
                extendedCrumbs: 'c'
              }
            ]
          }
        ]
      }
    ]

    const handleLocationChangeMock = jest.spyOn(handleLocationChange, 'default')
    let innerHandleLocationChange = null

    handleLocationChangeMock.mockImplementation(_handleLocationChange => {
      innerHandleLocationChange = _handleLocationChange
    })

    const { container } = render(
      <BreadcrumbsBanner
        crumbLookup={crumbsLookup}
        crumbResourceLookup={crumbResourceLookup}
        routeCrumbsTree={crumbs}
      />
    )
    act(() => {
      innerHandleLocationChange({ pathname: '/a/b/c' })
    })

    expect(container).toMatchSnapshot()
  })
})

describe('BreadcrumbsBanner behavior', () => {
  it('last crumb can not be click', () => {
    const crumbClickHandler = jest.fn()
    const crumbsLookup: IBreadcrumbsBannerProps['crumbLookup'] = {
      a: {},
      b: { onClick: crumbClickHandler }
    }
    const crumbResourceLookup: IBreadcrumbsBannerProps['crumbResourceLookup'] = {
      a: { name: 'a' },
      b: { name: 'b' }
    }
    const crumbs: IBreadcrumbsBannerProps['routeCrumbsTree'] = [
      {
        route: '/a',
        extendedCrumbs: 'a',
        children: [
          {
            route: '/a/b',
            extendedCrumbs: 'b'
          }
        ]
      }
    ]

    const handleLocationChangeMock = jest.spyOn(handleLocationChange, 'default')
    let innerHandleLocationChange = null

    handleLocationChangeMock.mockImplementation(_handleLocationChange => {
      innerHandleLocationChange = _handleLocationChange
    })

    render(
      <BreadcrumbsBanner
        crumbLookup={crumbsLookup}
        crumbResourceLookup={crumbResourceLookup}
        routeCrumbsTree={crumbs}
      />
    )
    act(() => {
      innerHandleLocationChange({ pathname: '/a/b' })
    })

    const lastNav = screen.getByText('b')
    fireEvent.click(lastNav)

    expect(crumbClickHandler).toBeCalledTimes(0)
  })

  it('crumb can be click if it is not the last one', () => {
    const crumbClickHandler = jest.fn()
    const crumbsLookup: IBreadcrumbsBannerProps['crumbLookup'] = {
      a: { onClick: crumbClickHandler },
      b: {}
    }
    const crumbResourceLookup: IBreadcrumbsBannerProps['crumbResourceLookup'] = {
      a: { name: 'a' },
      b: { name: 'b' }
    }
    const crumbs: IBreadcrumbsBannerProps['routeCrumbsTree'] = [
      {
        route: '/a',
        extendedCrumbs: 'a',
        children: [
          {
            route: '/a/b',
            extendedCrumbs: 'b'
          }
        ]
      }
    ]

    const handleLocationChangeMock = jest.spyOn(handleLocationChange, 'default')
    let innerHandleLocationChange = null

    handleLocationChangeMock.mockImplementation(_handleLocationChange => {
      innerHandleLocationChange = _handleLocationChange
    })

    render(
      <BreadcrumbsBanner
        crumbLookup={crumbsLookup}
        crumbResourceLookup={crumbResourceLookup}
        routeCrumbsTree={crumbs}
      />
    )
    act(() => {
      innerHandleLocationChange({ pathname: '/a/b' })
    })

    const lastNav = screen.getByText('a')
    fireEvent.click(lastNav)

    expect(crumbClickHandler).toBeCalledTimes(1)
  })
})
