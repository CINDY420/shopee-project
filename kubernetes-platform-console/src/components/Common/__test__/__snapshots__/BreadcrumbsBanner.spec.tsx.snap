// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbsBanner render UI result should display crumb text according to crumbResourceLookup configuration 1`] = `
<div>
  <div
    class="sc-gsTCUz canLKZ"
  >
    <nav>
      <div
        class="ant-breadcrumb sc-bdfBwQ cJPoQN"
      >
        <span>
          <span
            class="ant-breadcrumb-link"
          >
            <div
              class="sc-dlfnbm CZBcd"
            >
              <div
                class="sc-hKgILt gqwVAe"
              >
                a
              </div>
            </div>
          </span>
          <span
            class="ant-breadcrumb-separator"
          >
            /
          </span>
        </span>
      </div>
    </nav>
  </div>
</div>
`;

exports[`BreadcrumbsBanner render UI result should display crumb text with description as prefix according to crumbLookup configuration 1`] = `
<div>
  <div
    class="sc-gsTCUz canLKZ"
  >
    <nav>
      <div
        class="ant-breadcrumb sc-bdfBwQ cJPoQN"
      >
        <span>
          <span
            class="ant-breadcrumb-link"
          >
            <div
              class="sc-dlfnbm CZBcd"
            >
              <div
                class="sc-hKgILt gqwVAe"
              >
                A:
                a
              </div>
            </div>
          </span>
          <span
            class="ant-breadcrumb-separator"
          >
            /
          </span>
        </span>
      </div>
    </nav>
  </div>
</div>
`;

exports[`BreadcrumbsBanner render UI result should render children crumb according to routeCrumbsTree configuration 1`] = `
<div>
  <div
    class="sc-gsTCUz canLKZ"
  >
    <nav>
      <div
        class="ant-breadcrumb sc-bdfBwQ cJPoQN"
      >
        <span>
          <span
            class="ant-breadcrumb-link"
          >
            <div
              class="sc-dlfnbm CZBcd"
            >
              <div
                class="sc-hKgILt gqwVAe"
              >
                a
              </div>
            </div>
          </span>
          <span
            class="ant-breadcrumb-separator"
          >
            /
          </span>
        </span>
        <span>
          <span
            class="ant-breadcrumb-link"
          >
            <div
              class="sc-dlfnbm CZBcd"
            >
              <div
                class="sc-hKgILt gqwVAe"
              >
                b
              </div>
            </div>
          </span>
          <span
            class="ant-breadcrumb-separator"
          >
            /
          </span>
        </span>
        <span>
          <span
            class="ant-breadcrumb-link"
          >
            <div
              class="sc-dlfnbm CZBcd"
            >
              <div
                class="sc-hKgILt gqwVAe"
              >
                c
              </div>
            </div>
          </span>
          <span
            class="ant-breadcrumb-separator"
          >
            /
          </span>
        </span>
      </div>
    </nav>
  </div>
</div>
`;
